ast/term.go

```diff
@@ -2392,6 +2398,59 @@ func filterObject(o Value, filter Value) (Value, error) {
 	}
 }
 
+// NOTE(philipc): The only way to get an ObjectKeyIterator should be
+// from an Object. This ensures that the iterator can have implementation-
+// specific details internally, with no contracts except to the very
+// limited interface.
+type ObjectKeysIterator interface {
+	Next() (*Term, bool)
+	Value() *Term
+	Reset()
+	Copy() ObjectKeysIterator
+}
+
+type objectKeysIterator struct {
+	obj     *object
+	numKeys int
+	index   int
+}
+
+func newobjectKeysIterator(o *object) ObjectKeysIterator {
+	return &objectKeysIterator{
+		obj:     o,
+		numKeys: o.Len(),
+		index:   0,
+	}
+}
+
+func (oki *objectKeysIterator) Next() (*Term, bool) {
+	if oki.index == oki.numKeys || oki.numKeys == 0 {
+		return nil, false
+	}
+	oki.index++
+	return oki.obj.keys[oki.index-1].key, true
+}
+
+func (oki *objectKeysIterator) Value() *Term {
+	if oki.index == oki.numKeys || oki.numKeys == 0 {
+		return nil
+	}
+	return oki.obj.keys[oki.index].key
+}
+
+func (oki *objectKeysIterator) Reset() {
+	oki.index = 0
+}
```

Comment:
💭 Do we need `Reset()`? Without it, I'm imagining that the keys slice is all the state we'd need... 🤔

Reply:
I'm also looking to rip out the key slice and replace it with a linked-list or other data structure with better insertion performance. I put in the slice-based implementation for now, so that I could verify that I hadn't broken anything in the evaluator with the switch to using iterators instead of raw key indexes. :sweat_smile:

Reply:
I've removed `Reset()` for now, since I wound up not needing it during implementation. It'll be simple enough to add back if we actually need it later!